casper/src/main/scala/coop/rchain/casper/finality/Finalizer.scala

```diff
@@ -0,0 +1,123 @@
+package coop.rchain.casper.finality
+
+import cats.effect.{Concurrent, Sync}
+import cats.implicits._
+import coop.rchain.blockstorage.BlockStore
+import coop.rchain.blockstorage.dag.{BlockDagRepresentation, BlockDagStorage}
+import coop.rchain.blockstorage.deploy.DeployStorage
+import coop.rchain.blockstorage.syntax._
+import coop.rchain.casper.{PrettyPrinter, SafetyOracle}
+import coop.rchain.dag.DagOps
+import coop.rchain.models.BlockHash.BlockHash
+import coop.rchain.shared.Log
+import coop.rchain.shared.syntax._
+import fs2.Stream
+
+final case class HashWithFT(hash: BlockHash, faultTolerance: Float)
+
+final class Finalizer[F[_]: Sync: Log: Concurrent: BlockStore: BlockDagStorage: SafetyOracle: DeployStorage](
+    faultToleranceThreshold: Float
+) {
+  def run(
+      dag: BlockDagRepresentation[F],
+      lastFinalizedBlockHash: BlockHash,
+      removeBLockDeploysF: BlockHash => F[Unit],
+      recordFinalizedBlockF: BlockHash => F[Unit]
+  ): F[BlockHash] =
+    for {
+      children <- dag.children(lastFinalizedBlockHash).map(_.getOrElse(Set.empty[BlockHash]))
+      // calculate fault tolerances for children
+      hfts <- Stream
+               .emits(children.toList.map(c => computeFaultTolerance(dag, c)))
+               .parJoinProcBounded
+               .compile
+               .toList
+               .map(
+                 _.zip(children)
+                   .map { case (ft, hash) => HashWithFT(hash, ft) }
+               )
+      // children that are above fault tolerance threshold
+      finalizedChildren = hfts
+        .filter(_.faultTolerance > faultToleranceThreshold)
+        .sortBy(_.faultTolerance)
+
+      // compute finalized children parents that are not recorded as finalized yet
+      toFinalize <- Stream
+                     .emits(
+                       finalizedChildren
+                         .map(c => Stream.eval(dag.getNonFinalizedAncestors(c.hash)))
+                     )
+                     .parJoinProcBounded
+                     .compile
+                     .toList
+                     .map(_.reduce(_ ++ _))
+      _ <- Stream
+            .emits(
+              toFinalize.toList.map(
+                hash =>
+                  Stream.eval(
+                    removeBLockDeploysF(hash) >>
+                      recordFinalizedBlockF(hash)
+                  )
+              )
+            )
+            .parJoinProcBounded
+            .compile
+            .drain
```

Comment:
The whole mapping can be done inside Stream. Try not to allocate new memory for the whole structure you are starting with and do the transformation inside Stream where things are _loaded_ on demand (pull concept of fs2).
```scala
_ <- Stream
      .emits(toFinalize.toList)
      .parEvalMapProcBounded(hash => removeBLockDeploysF(hash) >> recordFinalizedBlockF(hash))
      .compile
      .drain
```